From 415d7098febfbc80b6239c0873837e94552de90b Mon Sep 17 00:00:00 2001
From: Matthias Gorzellik <matthias.gorzellik@gmail.com>
Date: Wed, 10 Jul 2024 15:57:32 +0200
Subject: [PATCH 1/2] winebus.sys: Fix rotation for angles < 90deg.

---
 dlls/winebus.sys/bus_sdl.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/winebus.sys/bus_sdl.c b/dlls/winebus.sys/bus_sdl.c
index 1c438eafc6a..7680473addd 100644
--- a/dlls/winebus.sys/bus_sdl.c
+++ b/dlls/winebus.sys/bus_sdl.c
@@ -682,7 +682,7 @@ static NTSTATUS sdl_device_physical_effect_update(struct unix_device *iface, BYT
     struct sdl_device *impl = impl_from_unix_device(iface);
     int id = impl->effect_ids[index];
     SDL_HapticEffect effect = {0};
-    UINT16 direction;
+    INT16 direction;
     NTSTATUS status;
 
     TRACE("iface %p, index %u, params %p.\n", iface, index, params);
@@ -693,6 +693,7 @@ static NTSTATUS sdl_device_physical_effect_update(struct unix_device *iface, BYT
     /* The first direction we get from PID is in polar coordinate space, so we need to
      * remove 90° to make it match SDL spherical coordinates. */
     direction = (params->direction[0] - 9000) % 36000;
+    if (direction < 0) direction += 36000;
 
     switch (params->effect_type)
     {
-- 
GitLab


From b32fa2364314bfa38910719e3fae7605b8db94e1 Mon Sep 17 00:00:00 2001
From: Matthias Gorzellik <matthias.gorzellik@gmail.com>
Date: Wed, 10 Jul 2024 16:34:02 +0200
Subject: [PATCH 2/2] winebus.sys: Align logical max of angles to physical max
 defined in dinput.

---
 dlls/winebus.sys/hid.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/winebus.sys/hid.c b/dlls/winebus.sys/hid.c
index d31718c6293..78efe6116ce 100644
--- a/dlls/winebus.sys/hid.c
+++ b/dlls/winebus.sys/hid.c
@@ -600,7 +600,7 @@ static BOOL hid_descriptor_add_set_periodic(struct unix_device *iface)
             UNIT(1, 0x14), /* Eng Rot:Angular Pos */
             UNIT_EXPONENT(1, -2),
             LOGICAL_MINIMUM(1, 0),
-            LOGICAL_MAXIMUM(4, 36000),
+            LOGICAL_MAXIMUM(4, 35900),
             REPORT_SIZE(1, 16),
             REPORT_COUNT(1, 1),
             OUTPUT(1, Data|Var|Abs),
@@ -960,7 +960,7 @@ BOOL hid_device_add_physical(struct unix_device *iface, USAGE *usages, USHORT co
                 UNIT(1, 0x14), /* Eng Rot:Angular Pos */
                 UNIT_EXPONENT(1, -2),
                 LOGICAL_MINIMUM(1, 0),
-                LOGICAL_MAXIMUM(4, 36000),
+                LOGICAL_MAXIMUM(4, 35900),
                 REPORT_SIZE(1, 16),
                 REPORT_COUNT(1, 2),
                 OUTPUT(1, Data|Var|Abs),
-- 
GitLab
